javascript/ql/src/semmle/javascript/frameworks/EventEmitter.qll

```diff
@@ -0,0 +1,167 @@
+  }
+
+  /**
+   * A registration of an event handler on a particular EventEmitter.
+   */
+  abstract class EventRegistration extends DataFlow::Node {
+    EventEmitter emitter;
+
+    /** Gets the EventEmitter that the event handler is registered on. */
+    final EventEmitter getEmitter() {
+      result = emitter
+    }
+
+    /** Gets the name of the channel if possible. */
+    abstract string getChannel();
+
+    /** Gets the `i`th parameter in the callback registered as the event handler. */
+    abstract DataFlow::Node getCallbackParameter(int i);
+    
+    /**
+     * Gets a value that is returned by the event handler to the `dispatch` where the event was dispatched. 
+     * The default implementation is that no value can be returned.
+     */
+    DataFlow::Node getAReturnedValue(EventDispatch dispatch) { none() }
+  }
+
+  /**
+   * A dispatch of an event on an EventEmitter.
+   */
+  abstract class EventDispatch extends DataFlow::Node {
+    EventEmitter emitter;
+
+    /** Gets the emitter that the event dispatch happens on. */    
+    final EventEmitter getEmitter() {
+      result = emitter
+    }
+    
+    /** Gets the name of the channel if possible. */
+    abstract string getChannel();
+
+    /** Gets the `i`th argument that is send to the event handler. */
+    abstract DataFlow::Node getDispatchedArgument(int i);
+    
+    /**
+     * Holds if this event dispatch can send an event to the given even registration. 
+     * The default implementation is that the emitters of the dispatch and registration has to be equal.
+     */
+    predicate canSendTo(EventRegistration destination) { this.getEmitter() = destination.getEmitter() }
+  }
+  
+  /**
+   * A taint-step that models data-flow between event handlers and event dispatchers.
+   */
+  private class EventEmitterTaintStep extends DataFlow::AdditionalFlowStep {
+    EventRegistration reg;
+    EventDispatch dispatch;
+
+    EventEmitterTaintStep() {
+      this = dispatch and 
+      dispatch.canSendTo(reg) and 
+      reg.getChannel() = dispatch.getChannel()
+    }
+
+    override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
+      exists(int i | i >= 0 | 
+        pred = dispatch.getDispatchedArgument(i) and
+        succ = reg.getCallbackParameter(i)
+      )
+      or
+      pred = reg.getAReturnedValue(dispatch) and
+      succ = dispatch
+    }
+  }
+
+  /**
+   * Concrete classes for modelling EventEmitter in NodeJS.
+   */
+  private module NodeJSEventEmitter {
+    private class NodeJSEventEmitter extends EventEmitter {
+      NodeJSEventEmitter() {
+        exists(DataFlow::SourceNode clazz |
+          clazz = DataFlow::moduleImport("events") or
+          clazz = DataFlow::moduleMember("events", "EventEmitter")
+        |
+          this = clazz.getAnInstantiation()
+        )
+      }
+    }
+
+    private class EventEmitterRegistration extends EventRegistration, DataFlow::MethodCallNode {
+      override EventEmitter emitter;
```

Comment:
This override doesn't actually sharpen the type of `emitter`. Should it?

Reply:
It used to sharpen the type in a previous iteration of the code, and I didn't remove that line after it became redundant.  

I'll remove the override.